wp-content/plugins/wporg-learn/inc/locale.php

```diff
@@ -156,3 +157,162 @@ function disable_caching( $headers ) {
+ * FROM wp_posts
+ * INNER JOIN wp_postmeta pmeta ON ( wp_posts.ID = pmeta.post_id )
+ * WHERE 1=1
+ * AND wp_posts.post_type = 'wporg_workshop'
+ * GROUP BY wp_posts.ID
+ * ORDER BY has_language DESC, has_caption DESC, wp_posts.post_date DESC
+ *
+ * @param array  $clauses
+ * @param string $locale
+ *
+ * @return array
+ */
+function wporg_tutorials_query_prioritize_locale( $clauses, $locale ) {
+	global $wpdb;
+
+	$locale_root = preg_replace( '#^([a-z]{2,3}_?)[a-zA-Z_-]*#', '$1', $locale, -1, $count );
+
+	if ( $count ) {
+		/**
+		 * $clauses['fields'] contains the SELECT part of the query.
+		 *
+		 * The extra fields clauses are calculated fields that will contain a `1` if the workshop post row has a postmeta
+		 * value that matches the locale root. The MAX() and the groupby clause below ensure that all the rows for a
+		 * given workshop are consolidated into one with the highest value in the calculated column. Without the
+		 * grouping, there would be a separate row for each postmeta value for each workshop post.
+		 */
+		$clauses['fields'] .= ",
+			MAX( IF( pmeta.meta_key = 'language' AND pmeta.meta_value LIKE '{$locale_root}%', 1, 0 ) ) AS has_language
+		";
+		$clauses['fields'] .= ",
+			MAX( IF( pmeta.meta_key = 'video_caption_language' AND pmeta.meta_value LIKE '{$locale_root}%', 1, 0 ) ) AS has_caption
+		";
+		$clauses['join']   .= " INNER JOIN {$wpdb->postmeta} pmeta ON ( {$wpdb->posts}.ID = pmeta.post_id )";
+		// This orderby clause ensures that the workshops are sorted by the values in the calculated columns first.
+		$clauses['orderby'] = 'has_language DESC, has_caption DESC, ' . $clauses['orderby'];
+
+		if ( false === strpos( $clauses['groupby'], "{$wpdb->posts}.ID" ) ) {
+			$clauses['groupby'] = "{$wpdb->posts}.ID";
+		}
+	}
+
+	return $clauses;
+}
+
+/**
+ * Modify the a post type archive query to prioritize posts in the user's locale.
+ *
+ * In order to show all posts, but with the ones that are presented in the user's locale shown first, we
+ * need to modify the posts query in ways that can't be done through the WP_Query or WP_Meta_Query APIs. Instead, here,
+ * we're filtering the individual clauses of the query to add the pieces we need.
+ *
+ * Examples, slightly truncated for simplicity:
+ *
+ * Before:
+ * SELECT SQL_CALC_FOUND_ROWS wp_posts.ID
+ * FROM wp_posts
+ * WHERE 1=1
+ * AND wp_posts.post_type = 'lesson'
+ * ORDER BY wp_posts.post_date DESC
+ *
+ * After:
+ * SELECT SQL_CALC_FOUND_ROWS wp_posts.*,
+ *   MAX( IF( pmeta.meta_key = 'language' AND pmeta.meta_value LIKE 'art_%', 1, 0 ) ) AS has_language,
+ * FROM wp_posts
+ * INNER JOIN wp_postmeta pmeta ON ( wp_posts.ID = pmeta.post_id )
+ * WHERE 1=1
+ * AND wp_posts.post_type = 'lesson'
+ * GROUP BY wp_posts.ID
+ * ORDER BY has_language DESC, wp_posts.post_date DESC
+ *
+ * @param array  $clauses
+ * @param string $locale
+ *
+ * @return array
+ */
+function wporg_query_prioritize_locale( $clauses, $locale ) {
```

Comment:
It seems that these two functions have some similarity. Do you think it would be clearer to combine them into one function like: `wporg_query_prioritize_locale($clauses, $locale, $post_type = '')`?

The condition inside could be changed to something like this:
```
// Common field for both post types
$clauses['fields'] .= ",
    MAX(
        CASE
            WHEN pmeta.meta_key = 'language' AND pmeta.meta_value LIKE '{$locale_root}%' THEN 1
            WHEN (pmeta.meta_key = 'language' AND pmeta.meta_value = '') OR pmeta.meta_key IS NULL THEN " . ( $is_english ? '1' : '0' ) . '
            ELSE 0
        END
    ) AS has_language
';

// Additional fields and joins for 'workshop' post type
if ( $post_type === 'workshop' ) {
    $clauses['fields'] .= ",
        MAX( IF( pmeta.meta_key = 'video_caption_language' AND pmeta.meta_value LIKE '{$locale_root}%', 1, 0 ) ) AS has_caption
    ";
}

$clauses['join'] .= " LEFT JOIN {$wpdb->postmeta} pmeta ON ( {$wpdb->posts}.ID = pmeta.post_id AND pmeta.meta_key IN ('language'" . ( $post_type === 'workshop' ? ", 'video_caption_language'" : "" ) . ") )";

$clauses['orderby'] = 'has_language DESC' . ( $post_type === 'workshop' ? ', has_caption DESC' : '' ) . ', ' . $clauses['orderby'];
$clauses['groupby'] = "{$wpdb->posts}.ID";
```
And the examples in the comments for the workshop can be moved here as well.

I feel that this way having the two query instructions together makes it clearer what the differences are. It would be faster to maintain and understand the code. This is just my initial thought when first seeing this part.

Reply:
This is a fair point. The main difference between the two is that the Tutorials one does not account for empty or null post meta being treated as English. That's the way prod is at present, and given Tutorials are being deprecated I thought I'd leave that query alone.

But we could improve it, what do you think?

Reply:
Maybe a deprecation note in the doc would make it clear that it has been superseded.

Reply:
A deprecation note makes sense 👍

Reply:
Done [6121cc3](https://github.com/WordPress/Learn/pull/2657/commits/6121cc338d9061bb7cc7716f4a310e155d5eaeff)